### ISSUE TITLE
🚀 Feature: AwsEKSClusterProcessor allow custom Entity Descriptor Callback

### ISSUE THREAD

**Issue by Sabrina0614** [backstage/backstage#18497]
### 🔖 Feature description

It would be great to have full control over the emitted EKS Cluster entities that are generated from the AwsEKSClusterProcessor.

This would be a fully backwards compatible and default to the currently generated EKS Cluster entity definitions.

### 🎤 Context

Doing so will allow for customization of emitted EKS Entities for Backstage instances that allows for just greater configuration of Entities based on the AWS cluster metadata.

This seemed like a potentially good approach to contribute back if y'all would find it useful to give the implementing catalog the ability to fine tune generated Entities.

Happy to open up a Pull Request as well for this!

### ✌️ Possible Implementation

For example, something like the following:
```diff
import {
  CatalogProcessor,
  CatalogProcessorEmit,
} from '@backstage/plugin-catalog-node';
import { LocationSpec } from '@backstage/plugin-catalog-common';
- import {
-  ANNOTATION_KUBERNETES_API_SERVER,
-  ANNOTATION_KUBERNETES_API_SERVER_CA,
-  ANNOTATION_KUBERNETES_AUTH_PROVIDER,
- } from '@backstage/plugin-kubernetes-common';
import { EKS } from '@aws-sdk/client-eks';
import { AWSCredentialFactory } from '../types';
import { AwsCredentialIdentity, Provider } from '@aws-sdk/types';
import {
  AwsCredentialsManager,
  DefaultAwsCredentialsManager,
} from '@backstage/integration-aws-node';
import { Config } from '@backstage/config';
+ import { defaultEKSClusterTransformer } from '../lib';
+ import type { Entity } from '@backstage/catalog-model';

- const ACCOUNTID_ANNOTATION: string = 'amazonaws.com/account-id';
- const ARN_ANNOTATION: string = 'amazonaws.com/arn';

+ export type EKSClusterEntityGenerator = (
+  cluster: Cluster,
+  accountID: string,
+ ) => Promise<Entity>;

/**
 * A processor for automatic discovery of resources from EKS clusters. Handles the
 * `aws-eks` location type, and target accounts/regions of the form
 * `<accountId>/<region>`.
 *
 * @public
 */
export class AwsEKSClusterProcessor implements CatalogProcessor {
  private credentialsFactory?: AWSCredentialFactory;
  private credentialsManager?: AwsCredentialsManager;
+ private readonly entityFromEKSCluster: EKSClusterEntityGenerator;

  static fromConfig(
    configRoot: Config,
+  options: {
+    entityFromEKSCluster?: EKSClusterEntityGenerator;
+   },
  ): AwsEKSClusterProcessor {
    const awsCredentaislManager =
      DefaultAwsCredentialsManager.fromConfig(configRoot);

    return new AwsEKSClusterProcessor({
      credentialsManager: awsCredentaislManager,
+   // Passing in the new paramter
+    ...options,
    });
  }

  constructor(options: {
    credentialsFactory?: AWSCredentialFactory;
    credentialsManager?: AwsCredentialsManager;
+  entityFromEKSCluster?: EKSClusterEntityGenerator;
  }) {
    this.credentialsFactory = options.credentialsFactory;
    this.credentialsManager = options.credentialsManager;
+  // If the callback function is not passed in, then default to the one upstream is using
+  this.entityFromEKSCluster =
+    options.entityFromEKSCluster || defaultEKSClusterTransformer;
  }

  getProcessorName(): string {
    return 'aws-eks';
  }

-  normalizeName(name: string): string {
-    return name
-      .trim()
-      .toLocaleLowerCase('en-US')
-      .replace(/[^a-zA-Z0-9\-]/g, '-');
-  }


  async readLocation(
    location: LocationSpec,
    _optional: boolean,
    emit: CatalogProcessorEmit,
  ): Promise<boolean> {
    if (location.type !== 'aws-eks') {
      return false;
    }

    // location target is of format "account-id/region"
    const [accountId, region] = location.target.split('/');

    if (!accountId || !region) {
      throw new Error(
        'AWS EKS location specified without account or region information',
      );
    }

    let credentials: AwsCredentialIdentity | undefined;

    if (this.credentialsFactory) {
      credentials = await this.credentialsFactory(accountId);
    }

    let providerFunction: (() => Provider<AwsCredentialIdentity>) | undefined;
    if (this.credentialsManager) {
      const credentialsProvider =
        await this.credentialsManager.getCredentialProvider({ accountId });
      providerFunction = () => credentialsProvider.sdkCredentialProvider;
    }

    const eksClient = new EKS({
      credentials,
      credentialDefaultProvider: providerFunction,
    });
    const clusters = await eksClient.listClusters({});
    if (clusters.clusters === undefined) {
      return true;
    }
-   const results = clusters.clusters
+   await Promise.all(
      clusters.clusters
        .map(cluster => eksClient.describeCluster({ name: cluster }))
        .map(async describedClusterPromise => {
          const describedCluster = await describedClusterPromise;
          describedCluster.cluster;
          if (describedCluster.cluster) {
+          // Call the callback function to generate the entity
+          const entity = await this.entityFromEKSCluster(
+            describedCluster.cluster,
+            accountId,
-          const entity = {
-            apiVersion: 'backstage.io/v1alpha1',
-            kind: 'Resource',
-            metadata: {
-             annotations: {
-                [ACCOUNTID_ANNOTATION]: accountId,
-                [ARN_ANNOTATION]: describedCluster.cluster.arn || '',
-                [ANNOTATION_KUBERNETES_API_SERVER]:
-                  describedCluster.cluster.endpoint || '',
-                [ANNOTATION_KUBERNETES_API_SERVER_CA]:
-                  describedCluster.cluster.certificateAuthority?.data || '',
-                [ANNOTATION_KUBERNETES_AUTH_PROVIDER]: 'aws',
-              },
-              name: this.normalizeName(describedCluster.cluster.name as string),
-              namespace: 'default',
-            },
-            spec: {
-              type: 'kubernetes-cluster',
-              owner: 'unknown',
-            },
-          };
            );

            emit({
              type: 'entity',
              entity,
              location,
            });
          }
        }),
    );
-  await Promise.all(results);
    return true;
  }
}
```

``` typescript
// plugins/catalog-backend-module-aws/src/lib/defaultTransformers.ts
import { EKSClusterEntityGenerator } from '../processors/types';
import { type Cluster } from '@aws-sdk/client-eks';
import {
  ANNOTATION_KUBERNETES_API_SERVER,
  ANNOTATION_KUBERNETES_API_SERVER_CA,
  ANNOTATION_KUBERNETES_AUTH_PROVIDER,
} from '@backstage/plugin-kubernetes-common';


const ACCOUNTID_ANNOTATION: string = 'amazonaws.com/account-id';
const ARN_ANNOTATION: string = 'amazonaws.com/arn';

/**
 * Default transformer for EKS Cluster to Resource Entity
 *
 * @public
 */
export const defaultEKSClusterTransformer: EKSClusterEntityGenerator = async (
  cluster: Cluster,
  accountId: string,
) => {
  const { arn, endpoint, certificateAuthority, name } = cluster;
  return {
    apiVersion: 'backstage.io/v1alpha1',
    kind: 'Resource',
    metadata: {
      annotations: {
        [ACCOUNTID_ANNOTATION]: accountId,
        [ARN_ANNOTATION]: arn || '',
        [ANNOTATION_KUBERNETES_API_SERVER]: endpoint || '',
        [ANNOTATION_KUBERNETES_API_SERVER_CA]: certificateAuthority?.data || '',
        [ANNOTATION_KUBERNETES_AUTH_PROVIDER]: 'aws',
      },
      name: normalizeName(name as string),
      namespace: 'default',
    },
    spec: {
      type: 'kubernetes-cluster',
      owner: 'unknown',
    },
  };
};

function normalizeName(name: string): string {
  return name
    .trim()
    .toLocaleLowerCase('en-US')
    .replace(/[^a-zA-Z0-9\-]/g, '-');
}
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by freben** [comment]
Sure! We have this concept in several processors / providers as you've seen. One tricky aspect is that this makes the AWS SDK types part of the public contract of the package, which potentially makes future upgrades a bit trickier since SDK dependency updates easily become unwittingly breaking for consumers. Can still be worth it, and can sometimes be mitigated by making a custom type with the important data fields copied out of the AWS type, if that's desirable.

**Comment by Sabrina0614** [comment]
Hello @freben, we noticed that the [region is required in the location entity](https://github.com/backstage/backstage/blob/v1.16.0/plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.ts#L89-L93). However, it is never being used. I think the intention for region is being used/passed in to [initialize the EKS client](https://github.com/backstage/backstage/blob/v1.16.0/plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.ts#L108-L111). Since region is not currently passed in when initializing the EKS client, backstage is only querying the region that the backstage is deploying to

Curious if you think it would be a great idea that instead of throwing the error if the region does not exist, the EKS client is default to the region that backstage is deployed to.

**Comment by Rugvip** [comment]
Closing as fixed by #19348
